src/roswire/ros2/reader.py

```diff
@@ -0,0 +1,141 @@
+        pass
+
+    def read(self,
+             filename: str,
+             argv: Optional[Sequence[str]] = None
+             ) -> LaunchConfig:
+        """Produces a summary of the effects of a launch file.
+
+           Parameters
+           ----------
+           filename: str
+               The name of the launch file, or an absolute path to the launch
+               file inside the container.
+           package: str, optional
+               The name of the package to which the launch file belongs.
+           argv: Sequence[str], optional
+               An optional sequence of command-line arguments that should be
+               supplied to :code:`roslaunch`.
+
+           Raises
+           ------
+           PackageNotFound
+               If the given package could not be found.
+           LaunchFileNotFound
+               If the given launch file could not be found in the package.
+           """
+
+        # Copy resources/launch_extractor._py as a python
+        # file into the container
+        logger.debug("Copying launch extraction script")
+        files = self.app_instance.files
+        host_script = pkg_resources. \
+            resource_filename('roswire',
+                              'resources/launch_extractor._py')
+        logger.debug(f'Resource is {host_script}')
+        files.copy_from_host(host_script,
+                             '/launch_extractor.py')
+
+        # Runs the script using app_instance.shell.
+        output = shlex.quote(os.path.basename(filename) + '.json')
+        cmd = f'python3 /launch_extractor.py --output' \
+              f' {output} {shlex.quote(filename)}'
+        logger.debug(f"Running the script in the container: {cmd}")
+        self.app_instance.shell.check_call(cmd)
+        logger.debug(f"Reading {output} on container")
+        config_json = files.read(output)
+        config_nodes = json.loads(config_json)
+        # Convert json to a launch config
+        # TODO Implement launch config
+        lc = LaunchConfig()
+        ctx = LaunchContext(namespace='/', filename=filename)
+        if argv:
+            ctx = ctx.with_argv(argv)
+
+        ctx, cfg = self._load_launch_objects(ctx, lc,
+                                             [list(config_nodes)])
+        logger.debug(f'launch configuration: {cfg}')
+        return lc
+
+    def _load_launch_objects(self, ctx: LaunchContext,
+                             cfg: LaunchConfig,
+                             node_list: Sequence[Sequence[Dict]]
+                             ) -> Tuple[LaunchContext, LaunchConfig]:
+        for nodes in node_list:
+            for node in nodes:
+                if node['__TYPE__'] == 'Node':
+                    print(f'processing: {node}')
+                    # remappings = self.\
+                    #     _get_remappings_from_json(node.get('remappings'))
+                    args: Optional[Any] = node.get('args', [])
```

Comment:
`Optional[Any]` seems like an overly strong type. Also, note that `args` will never be `None` here since if `args` is absent from `node`, `[]` will be used.

Reply:
Done